File: Python_code/Image_deconvolution/functions/plot_results.py
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import numpy as np
from IPython import get_ipython

# Function to plot the autocorrelation of the slowest component
def autocorr(x):
    result = np.correlate(x, x, mode='full')
    return result[result.size // 2:]

def plot_results(Y,X,nStagesSKROCK,meanSamples,logPiTrace,mseValues,\
                     slowComponent):

    plt.rcParams.update({
        "text.usetex": True,
        "font.family": "sans-serif",
        "font.sans-serif": ["Helvetica"]})

    'exec(%matplotlib inline)'

    # 1. PLOT ORIGINAL, OBSERVATIONS AND ESTIMATES

    # Plot the original image
    plt.figure(figsize=(16,8))
    plt.subplot(231)
    plt.gray()
    plt.imshow(X)
    plt.title("Original image")

    # Plot the noisy observation
    plt.subplot(232)
    plt.gray()
    plt.imshow(Y)
    plt.title("Blurred and noisy observation")

    # Plot the MMSE of x
    plt.subplot(233)
    plt.gray()
    plt.imshow(meanSamples)
    plt.title("MMSE estimate of $x$")
    
    # Plot the \log\pi trace of the samples
    plt.subplot(234)
    plt.plot(np.arange(1,nStagesSKROCK*len(logPiTrace),nStagesSKROCK),logPiTrace)
    plt.xscale('log')
    plt.xlabel('number of gradient evaluations')
    plt.ylabel('$\log\pi(X_n)$')
    plt.title('$\log\pi$ trace of $X_n$')
    
    # % Plot the evolution of the MSE in stationarity
    plt.subplot(235)
    plt.plot(np.arange(1,nStagesSKROCK*len(mseValues),nStagesSKROCK),mseValues)
    plt.xlabel('number of gradient evaluations')
    plt.ylabel('MSE')
    plt.title('Evolution of MSE in stationarity')
    
    # % Plot the autocorrelation function of the slowest component
    plt.subplot(236)
    lag  = int(2e2)
    autocorSKROCK = autocorr(slowComponent)
    autocorSKROCK = autocorSKROCK / autocorSKROCK.max()
    xVal = np.arange(0,lag+1,np.round(lag/20))
    xVal = xVal.astype(int)
    plt.stem(xVal,autocorSKROCK[xVal],markerfmt='^',use_line_collection=True)
    plt.xlabel('lag')
    plt.ylabel('ACF')
    plt.xlim(0,lag)
    plt.title('ACF')

    # to show all the plots
    plt.show()

File: MATLAB_code/Image_deconvolution/functions/plot_results.m
function plot_results(Y,X,nStagesSKROCK,meanSamples,logPiTrace,mseValues,...
                     slowComponent)

close all

% Dock the generated figures
set(0,'DefaultFigureWindowStyle','docked');

% Initial configurations for the plots
set(0, 'DefaultAxesFontSize',20);
set(0, 'DefaultLineLineWidth', 3);
set(0, 'defaultTextInterpreter', 'latex');

% 1. PLOT ORIGINAL, OBSERVATIONS AND ESTIMATES

    % Plot the original image
    figure(1);
    imagesc(X,[0 255]);
    title('Original image');
    axis equal; axis off;colormap('gray');

    % Plot the noisy observation
    figure(2);
    imagesc(Y, [0 255]); hold on
    axis equal; axis off;colormap('gray');
    title('Blurred and noisy observation');
 
    % Plot the MMSE of x
    figure(3);
    imagesc(meanSamples,[0 255]);
    axis equal; axis off;colormap('gray');
    title('MMSE estimate of x');
    
    % Plot the \log\pi trace of the samples
    figure(4);
    semilogx(1:nStagesSKROCK:nStagesSKROCK*length(logPiTrace),logPiTrace);
    xlabel('number of gradient evaluations');
    ylabel('$\log\pi(X_n)$','interpreter','latex');
    title('$\log\pi$ trace of $X_n$','Interpreter','latex');
    
    % Plot the evolution of the MSE in stationarity
    figure(5);
    semilogx(1:nStagesSKROCK:nStagesSKROCK*length(mseValues),mseValues);
    xlabel('number of gradient evaluations');
    ylabel('MSE','interpreter','latex');
    title('Evolution of MSE in stationarity','Interpreter','latex');
    
    % Plot the autocorrelation function of the slowest component
    figure(6);
    lag  = 20; 
    [autocorSKROCK,lags] = autocorr(slowComponent, lag);
    stem(lags,autocorSKROCK,'filled','^','LineWidth',2,'MarkerSize',10);
    legend({'SK-ROCK $(s=10)$'},'interpreter','latex');
    xlabel('lag','interpreter','latex');
    ylabel('ACF','interpreter','latex');
    xlim([0 lag]);
    title('ACF','Interpreter','latex');

end